src/levanter/layers/attention.py

```diff
@@ -220,6 +220,91 @@ def dot_product_attention(
         )
 
 
+def dot_product_attention_with_sink(
+    QPos: AxisSelector,
+    KPos: AxisSelection,
+    Key: AxisSelector,
+    query: NamedArray,
+    key: NamedArray,
+    value: NamedArray,
+    attn_sink: NamedArray,
+    mask: Optional[Union["AttentionMask", NamedArray]] = None,
+    bias: Optional[NamedArray] = None,
+    attention_dtype: Optional[jnp.dtype] = None,
+    precision: PrecisionLike = None,
+    use_flash: Optional[bool] = None,
+    attn_backend: Optional[AttentionBackend] = None,
+    flash_block_size: Optional[int] = None,
+    dropout: float = 0.0,
+    *,
+    logits_soft_cap: float | None = None,
+    scaling_factor: float | None = None,
+    inference: bool = True,
+    prng: PRNGKeyArray | None = None,
+):
+    """Dot-product attention variant with a learned sink term per head.
+
+    The sink is implemented by appending a dummy key/value of zeros and
+    inserting the sink logit via the bias term at the final key position.
+    """
+
+    QPos = query.resolve_axis(QPos)
+    KPos = key.resolve_axis(KPos)
+    Key = query.resolve_axis(Key)
+
+    KPos1 = KPos.resize(1)
+    KPosPlus = KPos.resize(KPos.size + 1)
+
+    zero_key_axes = tuple(KPos1 if ax == KPos else ax for ax in key.axes)
+    zero_key = hax.zeros(zero_key_axes, dtype=key.dtype)
+    key = hax.concatenate(KPosPlus, [key, zero_key])
+
+    zero_val_axes = tuple(KPos1 if ax == KPos else ax for ax in value.axes)
+    zero_val = hax.zeros(zero_val_axes, dtype=value.dtype)
+    value = hax.concatenate(KPosPlus, [value, zero_val])
+
+    m = materialize_mask(mask, QPos, KPos)
+    if m is not None:
+        sink_mask_axes = tuple(KPos1 if ax == KPos else ax for ax in m.axes)
+        sink_mask = hax.ones(sink_mask_axes, dtype=m.dtype)
+        m = hax.concatenate(KPosPlus, [m, sink_mask])
+
+    bias_axes_prefix = tuple(ax for ax in query.axes if ax != Key)
+    sink_bias = attn_sink
+    for ax in bias_axes_prefix:
+        if ax not in sink_bias.axes:
+            sink_bias = sink_bias.broadcast_axis(ax)
+    sink_bias = sink_bias.broadcast_axis(KPos1)
+
+    if bias is not None:
+        bias = hax.concatenate(KPosPlus, [bias, sink_bias])
+    else:
+        zero_bias_axes = bias_axes_prefix + (KPos,)
+        zero_bias = hax.zeros(zero_bias_axes, dtype=sink_bias.dtype)
+        bias = hax.concatenate(KPosPlus, [zero_bias, sink_bias])
+
+    return dot_product_attention(
```

Comment:
my guess is the correct thing to do is to get the logz from attention and re-weight attention outputs with that?

Reply:
Overall yes, but I did reason I have to get into the low-level impl. of both Splash Attention (for TPU) and Flash Attention (for GPU) right unless I'm missing something.

Reply:
Going to go ahead and merge for now, we can come back to this later if attention sinks become more core in our usage.